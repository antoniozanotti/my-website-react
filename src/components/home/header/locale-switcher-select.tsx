"use client";

import { useRouter } from "next/navigation";
import { ChangeEvent, ReactNode, useTransition } from "react";

type Props = {
  children: ReactNode;
  defaultValue: string;
  label: string;
};

export function LocaleSwitcherSelect({ children, defaultValue, label }: Props) {
  const router = useRouter();
  const [isPending, startTransition] = useTransition();

  function onSelectChange(event: ChangeEvent<HTMLSelectElement>) {
    const nextLocale = event.target.value;
    startTransition(() => {
      router.push("/"+nextLocale);
    });
  }

  let labelClasses = "relative relative flex justify-around";
  if (isPending) {
    labelClasses += " transition-opacity [&:disabled]:opacity-30";
  }

  return (
    <label className={labelClasses}>
      <p className="sr-only">{label}</p>
      <select
        className="inline-flex appearance-none bg-secondary dark:bg-dark-secondary w-[153px] lg:w-[173px] h-[38px] lg:h-[48px] rounded-full cursor-pointer py-0 pl-4 pr-6"
        defaultValue={defaultValue}
        disabled={isPending}
        onChange={onSelectChange}
      >
        {children}
      </select>
      <span className="pointer-events-none absolute right-4 top-[6px] lg:top-[8px]">
        âŒ„
      </span>
    </label>
  );
}
