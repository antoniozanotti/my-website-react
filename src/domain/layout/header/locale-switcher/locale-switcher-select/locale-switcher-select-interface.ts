export interface LocaleSwitcherSelectInterface {
  defaultValue: string;
  label: string;
  children: React.ReactNode;
};