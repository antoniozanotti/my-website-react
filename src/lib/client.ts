import { ApolloClient, HttpLink, InMemoryCache } from "@apollo/client";
import { registerApolloClient } from "@apollo/experimental-nextjs-app-support/rsc";

const TOKEN = process.env.CONTENTFUL_ACCESS_TOKEN;
const SPACE = process.env.CONTENTFUL_SPACE_ID;
const ENVIRONMENT = process.env.CONTENTFUL_ENVIRONMENT_ID;
const URL = `https://graphql.contentful.com/content/v1/spaces/${SPACE}/environments/${ENVIRONMENT}`;

export const { getClient } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: URL,
      headers: {
        Authorization: `Bearer ${TOKEN}`,
      },
    }),
  });
});
