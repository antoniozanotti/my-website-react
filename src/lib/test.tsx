import React, { ReactElement } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";
import { Theme } from "@/domain/theme/theme/theme";
import { Providers } from "../app/providers";
import messages from "@/messages/en.json";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return (
    <Providers locale="en" messages={messages}>
      <Theme>
        <div data-testid="container">{children}</div>
      </Theme>
    </Providers>
  );
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
  };
});
