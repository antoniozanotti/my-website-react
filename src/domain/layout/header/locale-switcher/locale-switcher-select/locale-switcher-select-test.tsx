import { test, expect } from "@playwright/experimental-ct-react";
import { LocaleSwitcherSelect } from "./locale-switcher-select";

test.describe("LocaleSwitcher", () => {
  test("should render component", async ({ mount, page }) => {  
    const component = await mount(
      <LocaleSwitcherSelect defaultValue="en" label="Change Language">
        <option value="en">English</option>
        <option value="pt">PortuguÃªs</option>
      </LocaleSwitcherSelect>
    );
    await expect(component.getByRole("option")).toHaveCount(2);    
  });
});
