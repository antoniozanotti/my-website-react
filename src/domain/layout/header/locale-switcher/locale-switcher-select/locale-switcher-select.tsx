"use client";

import { LocaleSwitcherSelectInterface } from "./locale-switcher-select-interface";
import { useRouter } from "next/navigation";
import { ChangeEvent, useTransition } from "react";

export function LocaleSwitcherSelect({
  defaultValue,
  label,
  children,
}: LocaleSwitcherSelectInterface) {
  const router = useRouter();
  const [isPending, startTransition] = useTransition();

  function onSelectChange(event: ChangeEvent<HTMLSelectElement>) {
    const nextLocale = event.target.value;
    startTransition(() => {
      router.push("/" + nextLocale);
    });
  }

  const labelBaseStyles = "relative relative flex justify-around";
  const labelIsPendingStyles = isPending
    ? "transition-opacity [&:disabled]:opacity-30"
    : "";
  const labelStyles = `${labelBaseStyles} ${labelIsPendingStyles}`;

  return (
    <label className={labelStyles}>
      <p className="sr-only">{label}</p>
      <select
        className="inline-flex appearance-none bg-secondary dark:bg-dark-secondary w-[153px] lg:w-[173px] h-[38px] lg:h-[48px] rounded-full cursor-pointer py-0 pl-4 pr-6"
        defaultValue={defaultValue}
        disabled={isPending}
        onChange={onSelectChange}
      >
        {children}
      </select>
      <span className="pointer-events-none absolute right-4 top-[6px] lg:top-[8px]">
        âŒ„
      </span>
    </label>
  );
}
