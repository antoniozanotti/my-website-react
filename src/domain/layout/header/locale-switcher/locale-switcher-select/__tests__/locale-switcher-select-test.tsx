import { screen, render } from "@/lib/test";
import { LocaleSwitcherSelect } from "../locale-switcher-select";

describe("LocaleSwitcher", () => {
  it("should render component", async () => {
    const { container } = render(
      <LocaleSwitcherSelect defaultValue="en" label="Change Language">
        <option value="en">English</option>
        <option value="pt">PortuguÃªs</option>
      </LocaleSwitcherSelect>
    );
    
    expect(screen.getAllByRole("option")).toHaveLength(2);
    expect(container).toMatchSnapshot();
  });
});
