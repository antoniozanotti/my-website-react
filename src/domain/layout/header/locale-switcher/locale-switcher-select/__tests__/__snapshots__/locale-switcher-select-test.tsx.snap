// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleSwitcher should render component 1`] = `
<div>
  <div
    class="dark"
  >
    <div
      class="className bg-dark-2 dark:bg-dark-dark-2 text-primary dark:text-dark-primary text-c3 lg:text-c4"
    >
      <div
        data-testid="container"
      >
        <label
          class="relative relative flex justify-around "
        >
          <p
            class="sr-only"
          >
            Change Language
          </p>
          <select
            class="inline-flex appearance-none bg-secondary dark:bg-dark-secondary w-[153px] lg:w-[173px] h-[38px] lg:h-[48px] rounded-full cursor-pointer py-0 pl-4 pr-6"
          >
            <option
              selected=""
              value="en"
            >
              English
            </option>
            <option
              value="pt"
            >
              Português
            </option>
          </select>
          <span
            class="pointer-events-none absolute right-4 top-[6px] lg:top-[8px]"
          >
            ⌄
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
